# [Day 1-3] 프로젝트 설명서

---

| 아래의 링크를 통해 파일을 다운로드 받으세요.  
| https://ds-lecture-data.s3.ap-northeast-2.amazonaws.com/datasets/vgames2.csv

---

# 목표 
+ 다음 분기에 어떤 게임을 설계해야 할까

---

# 🕹️ Data Description
+ Name : 게임의 이름입니다.  
+ Platform : 게임이 지원되는 플랫폼의 이름입니다.  
+ Year : 게임이 출시된 연도입니다.  
+ Genre : 게임의 장르입니다.  
+ Publisher : 게임을 제작한 회사입니다.  
+ NA_Sales : 북미지역에서의 출고량입니다.  
+ EU_Sales : 유럽지역에서의 출고량입니다.  
+ JP_Sales : 일본지역에서의 출고량입니다.  
+ Other_Sales : 기타지역에서의 출고량입니다.  

---

# 참조  
+ Sprint 1 : EDA, Feature Engineering, Data Manipulation, Data Visualization  
+ Sprint 2 : Hypothesis Test, CI, Bayesian, AB Test  
+ Sprint 3 : DR, PCA, Clustering, GD  

---

# 반드시 들어가야 하는 내용.
+ 지역에 따라서 선호하는 게임 장르가 다를까
+ 연도별 게임의 트렌드가 있을까
+ 출고량이 높은 게임에 대한 분석 및 시각화 프로세스가 포함

# 라이브러리 임포트

In [236]:
# 데이터 핸들링
import pandas as pd
import numpy as np

# 시각화
import matplotlib.pyplot as plt
import seaborn as sns

# 데이터 로드

In [237]:
# data from github public repository
# 기존 데이터에 인덱스가 포함되어 있음으로, 인덱스 없이 불러오기. https://jkim83.tistory.com/169
df_origin = pd.read_csv("https://github.com/CJYjob/ds-section1-Project1-new-/blob/main/vgames2.csv?raw=true", index_col=0)

# 데이터 원본 보존
원본 = df_origin  
가공 = df

In [238]:
# 데이터 원본 보존
df = df_origin.copy()

# Data Wrangling : 탐색

In [239]:
df.shape

(16598, 9)

In [240]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 16598 entries, 1 to 16598
Data columns (total 9 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   Name         16598 non-null  object 
 1   Platform     16598 non-null  object 
 2   Year         16327 non-null  float64
 3   Genre        16548 non-null  object 
 4   Publisher    16540 non-null  object 
 5   NA_Sales     16598 non-null  object 
 6   EU_Sales     16598 non-null  object 
 7   JP_Sales     16598 non-null  object 
 8   Other_Sales  16598 non-null  object 
dtypes: float64(1), object(8)
memory usage: 1.3+ MB


In [241]:
df.head()

Name Platform    Year      Genre  \
1                    Candace Kane's Candy Factory       DS  2008.0     Action   
2                                  The Munchables      Wii  2009.0     Action   
3  Otome wa Oanesama Boku ni Koi Shiteru Portable      PSP  2010.0  Adventure   
4                Deal or No Deal: Special Edition       DS  2010.0       Misc   
5       Ben 10 Ultimate Alien: Cosmic Destruction      PS3  2010.0   Platform   

            Publisher NA_Sales EU_Sales JP_Sales Other_Sales  
1           Destineer     0.04        0        0           0  
2  Namco Bandai Games     0.17        0        0        0.01  
3           Alchemist        0        0     0.02           0  
4           Zoo Games     0.04        0        0           0  
5         D3Publisher     0.12     0.09        0        0.04

In [242]:
df.tail()

Name Platform    Year     Genre  \
16594      Ice Age 2: The Meltdown       GC  2006.0  Platform   
16595  Rainbow Islands: Revolution      PSP  2005.0    Action   
16596                     NBA 2K16      PS3  2015.0    Sports   
16597  Toukiden: The Age of Demons      PSV  2013.0    Action   
16598     The King of Fighters '95       PS  1996.0  Fighting   

                         Publisher NA_Sales EU_Sales JP_Sales Other_Sales  
16594                Vivendi Games     0.15     0.04        0        0.01  
16595            Rising Star Games     0.01        0        0           0  
16596         Take-Two Interactive     0.44     0.19     0.03        0.13  
16597                   Tecmo Koei     0.05     0.05     0.25        0.03  
16598  Sony Computer Entertainment        0        0     0.16        0.01

# Data Wrangling : 탐색

In [243]:
df.Platform.unique()

array(['DS', 'Wii', 'PSP', 'PS3', 'PC', 'PS', 'GBA', 'PS4', 'PS2', 'XB',
       'X360', 'GC', '3DS', '2600', 'SAT', 'GB', 'NES', 'DC', 'N64',
       'XOne', 'SNES', 'WiiU', 'PSV', 'GEN', 'SCD', 'WS', 'NG', 'TG16',
       '3DO', 'GG', 'PCFX'], dtype=object)

In [244]:
df.Genre.unique()

array(['Action', 'Adventure', 'Misc', 'Platform', 'Sports', 'Simulation',
       'Racing', 'Role-Playing', 'Puzzle', 'Strategy', 'Fighting',
       'Shooter', nan], dtype=object)

In [245]:
df.NA_Sales.unique()

array(['0.04', '0.17', '0', '0.12', '0.02', '0.07', '480K', '0.14',
       '0.01', '0.57', '0.06', '0.08', '0.48', '0.03', '0.05', '1.93',
       '0.29', '0.19', '0.3', '0.26', '0.11', '0.18', '0.33', '0.51',
       '60K', '0.13', '0.52', '0.34', '0.25', '0.41', '0.22', '0.35',
       '0.1', '0.67', '0.59', '0.36', '0.21', '1.19', '5.91', '1.98',
       '0.09', '0K', '0.5', '0.94', '0.16', '0.63', '0.43', '0.2', '1.23',
       '0.44', '0.42', '0.97', '2.67', '0.28', '6.05', '0.47', '0.49',
       '0.62', '0.24', '0.72', '1.74', '0.46', '0.65', '0.32', '1.53',
       '0.23', '0.53', '0.37', '1.26', '0.39', '0.58', '1.94', '2.31',
       '1.03', '0.82', '0.4', '0.93', '0.56', '0.96', '1.64', '0.69',
       '1.25', '0.27', '0.92', '0.58M', '1.24', '0.25M', '0.86', '0.73',
       '0.55', '1.54', '1.02', '1.27', '0.54', '0.15', '1.09', '3.44',
       '2.29', '0.88', '5.55M', '1.37', '2.3', '0.99', '1.15', '0.98',
       '0.75', '0.74', '0.7', '0.14M', '0.45', '1.2', '1.41', '0.68',
       '

# Data Wrangling : 문제 정의

+ 총 16598개의 관측치가 있으며, 
  + Year column에 271 개의 결측치,
  + Genre column에 50 개의 결측치
  + Publisher column에 58 개의 결측치  
  가 관찰됨.
+ 이산형 데이터인 year가 float형으로 설정되어 있음. -> int형으로 전환
+ 판매량이 object형이며 세가지 형태(420K 0.28M 0.01)로 표현 -> int형으로 전환
+ 구조적으로 Tidy하지 못한 데이터(	NA_Sales	EU_Sales	JP_Sales	Other_Sales) -> Location,Sales 칼럼으로 전환

In [246]:
# Year의 NaN 데이터 개수 확인 
year_NaN_count = df.Year.isna().sum()
genre_NaN_count = df.Genre.isna().sum()
publisher_NaN_count = df.Publisher.isna().sum()
print(year_NaN_count, genre_NaN_count, publisher_NaN_count)

271 50 58


In [247]:
# 연도를 추측할 방법이 없음으로, drop 고려
# 결측치 비율 계산
ratio = (year_NaN_count+genre_NaN_count+publisher_NaN_count) / df.shape[0] * 100
print(ratio)

2.2834076394746354


In [248]:
# 결측치가 모두 다른 관측치에서 발견된다 가정해도, 
# 전체 중 2.28%로 5% 미만 임으로 결측치로 제거하여도 큰 영향이 없음.
df = df.dropna(axis=0)
df.isna().sum()

Name           0
Platform       0
Year           0
Genre          0
Publisher      0
NA_Sales       0
EU_Sales       0
JP_Sales       0
Other_Sales    0
dtype: int64

In [249]:
# df.year 의 데이터 타입 전환
df = df.astype({'Year':'int'})

In [250]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 16241 entries, 1 to 16598
Data columns (total 9 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   Name         16241 non-null  object
 1   Platform     16241 non-null  object
 2   Year         16241 non-null  int64 
 3   Genre        16241 non-null  object
 4   Publisher    16241 non-null  object
 5   NA_Sales     16241 non-null  object
 6   EU_Sales     16241 non-null  object
 7   JP_Sales     16241 non-null  object
 8   Other_Sales  16241 non-null  object
dtypes: int64(1), object(8)
memory usage: 1.2+ MB


In [251]:
# tidy data로 전환
df = df.melt(id_vars=['Name', 'Platform', 'Year', 'Genre', 'Publisher'], value_vars=['NA_Sales', 'EU_Sales', 'JP_Sales', 'Other_Sales'])
df = df.rename(columns = {'variable' :'Location', 'value':'Sales'})    
df["Location"] = df["Location"].str.replace("_Sales", "")

In [252]:
df

Name Platform  Year  \
0                        Candace Kane's Candy Factory       DS  2008   
1                                      The Munchables      Wii  2009   
2      Otome wa Oanesama Boku ni Koi Shiteru Portable      PSP  2010   
3                    Deal or No Deal: Special Edition       DS  2010   
4           Ben 10 Ultimate Alien: Cosmic Destruction      PS3  2010   
...                                               ...      ...   ...   
64959                         Ice Age 2: The Meltdown       GC  2006   
64960                     Rainbow Islands: Revolution      PSP  2005   
64961                                        NBA 2K16      PS3  2015   
64962                     Toukiden: The Age of Demons      PSV  2013   
64963                        The King of Fighters '95       PS  1996   

           Genre                    Publisher Location Sales  
0         Action                    Destineer       NA  0.04  
1         Action           Namco Bandai Games       NA  0.17  
2      Adventure                    Alchemist       NA     0  
3           Misc                    Zoo Games       NA  0.04  
4       Platform                  D3Publisher       NA  0.12  
...          ...                          ...      ...   ...  
64959   Platform                Vivendi Games    Other  0.01  
64960     Action            Rising Star Games    Other     0  
64961     Sports         Take-Two Interactive    Other  0.13  
64962     Action                   Tecmo Koei    Other  0.03  
64963   Fighting  Sony Computer Entertainment    Other  0.01  

[64964 rows x 7 columns]

In [253]:
# 판매량이 세가지 형태로 표현. x.xx, x.xxM, xxxK
# Sales에 대해서 세가지 형태(문자 포함 x, M 포함, K 포함)로만 칼럼이 이루어 지는지 확인
df.shape[0] == df[~df.Sales.str.contains("M")&~df.Sales.str.contains("K")]["Sales"].count() + df[df.Sales.str.contains("M")]["Sales"].count() + df[df.Sales.str.contains("K")]["Sales"].count()

True

In [254]:
# scale을 표시하기 위한 column 추가
df["Scale"] = np.repeat(0, len(df))

In [255]:
for i in df[df["Sales"].str.contains("K")].index:
  df.loc[i, "Scale"] = 1000
for i in df[~df["Sales"].str.contains("K")].index:
  df.loc[i, "Scale"] = 1000000

In [256]:
# 판매량을 수치형으로 표현하기 위해, "M","K" 제거
df["Sales"] = df["Sales"].str.replace("M", "")
df["Sales"] = df["Sales"].str.replace("K", "")

In [257]:
# 판매량 계산을 위해 float형으로 전환
df["Sales"] = df["Sales"].astype(float)

In [258]:
# 판매량 계산 후 업데이트
df["Sales"] = df["Sales"] * df["Scale"]

In [259]:
# 판매량을 int형으로 전환
df["Sales"] = df["Sales"].astype(int)

In [261]:
# 계산을 마친 Scale 칼럼 버리기
df = df.iloc[:,:-1]

In [262]:
df

Name Platform  Year  \
0                        Candace Kane's Candy Factory       DS  2008   
1                                      The Munchables      Wii  2009   
2      Otome wa Oanesama Boku ni Koi Shiteru Portable      PSP  2010   
3                    Deal or No Deal: Special Edition       DS  2010   
4           Ben 10 Ultimate Alien: Cosmic Destruction      PS3  2010   
...                                               ...      ...   ...   
64959                         Ice Age 2: The Meltdown       GC  2006   
64960                     Rainbow Islands: Revolution      PSP  2005   
64961                                        NBA 2K16      PS3  2015   
64962                     Toukiden: The Age of Demons      PSV  2013   
64963                        The King of Fighters '95       PS  1996   

           Genre                    Publisher Location   Sales  
0         Action                    Destineer       NA   40000  
1         Action           Namco Bandai Games       NA  170000  
2      Adventure                    Alchemist       NA       0  
3           Misc                    Zoo Games       NA   40000  
4       Platform                  D3Publisher       NA  120000  
...          ...                          ...      ...     ...  
64959   Platform                Vivendi Games    Other   10000  
64960     Action            Rising Star Games    Other       0  
64961     Sports         Take-Two Interactive    Other  130000  
64962     Action                   Tecmo Koei    Other   30000  
64963   Fighting  Sony Computer Entertainment    Other   10000  

[64964 rows x 7 columns]

In [263]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 64964 entries, 0 to 64963
Data columns (total 7 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   Name       64964 non-null  object
 1   Platform   64964 non-null  object
 2   Year       64964 non-null  int64 
 3   Genre      64964 non-null  object
 4   Publisher  64964 non-null  object
 5   Location   64964 non-null  object
 6   Sales      64964 non-null  int64 
dtypes: int64(2), object(5)
memory usage: 3.5+ MB
